tools/crates-validator/src/lib.rs

```diff
@@ -0,0 +1,562 @@
+        );
+        if profile.crossbeam_patch {
+            crate_line += CROSSBEAM_PATCH;
+        }
+
+        vars.insert("risc0_feature_std", risc0_feature_std);
+        vars.insert("crate_line", &crate_line);
+        vars.insert("risc0_zkvm", &risc0_zkvm);
+
+        Self::render_file(&cargo_toml_file, CARGO_TOML_TEMPLATE, vars)?;
+
+        // Generate the main.rs file
+
+        let mut vars = BTreeMap::new();
+        vars.insert("no_std_line", if profile.std { "" } else { "#![no_std]" });
+        vars.insert(
+            "use_lines",
+            if let Some(import_str) = &profile.import_str {
+                import_str
+            } else {
+                ""
+            },
+        );
+        vars.insert(
+            "main_body",
+            if let Some(main_body) = &profile.custom_main {
+                main_body
+            } else {
+                ""
+            },
+        );
+
+        Self::render_file(&main_file, MAIN_RS_TEMPLATE, vars)?;
+
+        Ok(())
+    }
+
+    // Builds the template project
+    fn build_project(&self, profile: &CrateProfile, working_dir: &Path) -> Result<(bool, String)> {
+        debug!(
+            "building {} - {}",
+            profile.name,
+            working_dir.to_string_lossy()
+        );
+
+        // NOTE:
+        // if you invoke this tool via `cargo run ...` it sets the CARGO_* and
+        // RUST_TOOLCHAIN env vars, here we strip out all the extra env vars so
+        // the cargo subcommands will correctly dispatch to rustup to pick the
+        // target project's rust-toolchain.toml.
+        let mut filtered_env: std::collections::HashMap<String, String> = std::env::vars()
+            .filter(|&(ref k, _)| {
+                k == "TERM" || k == "TZ" || k == "CC" || k == "CXX" || k == "LANG" || k == "PATH"
+            })
+            .collect();
+
+        // TODO: support GCC
+        if profile.inject_cc_flags {
+            filtered_env.insert("CC".to_string(), "clang".to_string());
+            filtered_env.insert(
+                "CFLAGS_riscv32im_risc0_zkvm_elf".to_string(),
+                "-target riscv32-unknown-elf".to_string(),
+            );
+        }
+
+        let guest_log_file = tempfile::NamedTempFile::new()?;
+        filtered_env.insert(
+            "RISC0_GUEST_LOGFILE".to_string(),
+            guest_log_file.path().to_string_lossy().to_string(),
+        );
+
+        let output = Command::new("cargo")
+            .args(["build"])
+            .env_clear()
+            .envs(&filtered_env)
+            .current_dir(working_dir)
+            .output()?;
+
+        let status = output.status;
+        let res = if status.success() || (!status.success() && profile.should_fail) {
+            info!("{} - build - SUCCESS", profile.name);
+            (true, String::new())
+        } else {
+            // Write out the build log to stderr
+            // TODO: Capture and analyze this log for common classes of errors?
```

Comment:
Definitely would be cool and helpful! And also until we can resource such a feature this can be done by manually looking at all the failures.

Reply:
Yeah my current thinking is to actually put that all into the UI code, since we can scan the snipped build logs in javascript. But maybe move it over here later, idk.